README.md
# Laravel package for ULID (Universally Unique Lexicographically Sortable Identifier)

[![Latest Version on Packagist](https://img.shields.io/packagist/v/vanthao03596/laravel-ulid.svg?style=flat-square)](https://packagist.org/packages/vanthao03596/laravel-ulid)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/vanthao03596/laravel-ulid/run-tests?label=tests)](https://github.com/vanthao03596/laravel-ulid/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/vanthao03596/laravel-ulid/Check%20&%20fix%20styling?label=code%20style)](https://github.com/vanthao03596/laravel-ulid/actions?query=workflow%3A"Check+%26+fix+styling"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/vanthao03596/laravel-ulid.svg?style=flat-square)](https://packagist.org/packages/vanthao03596/laravel-ulid)

## Installation

You can install the package via composer:

```bash
composer require vanthao03596/laravel-ulid
```

## Usage

Generate uld from Str support class

```php
// Default ulid generator with the current timestamp & lowercase string
Str::ulid(); // 01FDRXQ57VR4K4RASPT9NPAQC0
// Default ulid generator with the current timestamp & uppercase string
Str::ulid(false); // 01fdrxpmg9njp20z3km461fgax
// Default ulid generator with the given datetime & uppercase string
Str::ulid(false, Carbon::now()->subHour()) // 01FDRTD2K8Z664S24X606N14KD
```

Simply declare a ulid column type in your migration files.

```php
Schema::create('foos', function (Blueprint $table) {
    $table->ulid('id')->primary(); // adds primary ulid column 
    $table->foreignUlid('user_id')->constrained()->cascadeOnDelete(); // adds ulid foreignkey
    $table->ulidMorphs('taggable'); // adds ulid morphs
    $table->nullableUlidMorphs('likeable'); // adds nullable ulid morphs
});
```

Auto generate ulid column in your model

```php
<?php

namespace App;

use Illuminate\Database\Eloquent\Model;
use Vanthao03596\LaravelUlid\GeneratesUlid;

class User extends Model
{
    use GeneratesUlid;
}

```

Default column is `ulid`. If you wish to use a custom column name, for example if you want your primary `custom_column` column to be a `ULID`, you can define a `ulidColumn` method in your model.

```php
class User extends Model
{
    public function ulidColumn(): string
    {
        return 'custom_column';
    }
}
```

You can have multiple ULID columns in each table by specifying an array in the `ulidColumns` method. 

```php
class User extends Model
{
    public function ulidColumns(): array
    {
        return ['ulid', 'custom_column'];
    }
}
```
## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](.github/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [phamthao](https://github.com/vanthao03596)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `laravel-ulid` will be documented in this file.

## 0.1 - 2021-24-08

- initial release


tests/TestCase.php
<?php

namespace Vanthao03596\LaravelUlid\Tests;

use Illuminate\Database\Eloquent\Factories\Factory;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;
use Orchestra\Testbench\TestCase as Orchestra;
use Vanthao03596\LaravelUlid\LaravelUlidServiceProvider;

class TestCase extends Orchestra
{
    public function setUp(): void
    {
        parent::setUp();

        Factory::guessFactoryNamesUsing(
            fn (string $modelName) => 'Vanthao03596\\LaravelUlid\\Database\\Factories\\'.class_basename($modelName).'Factory'
        );

        $this->setupDatabase($this->app);
    }

    protected function getPackageProviders($app)
    {
        return [
            LaravelUlidServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app)
    {
        config()->set('database.default', 'testing');
    }

    protected function setupDatabase($app)
    {
        Schema::dropAllTables();

        $app['db']->connection()->getSchemaBuilder()->create('posts', function (Blueprint $table) {
            $table->increments('id');
            $table->ulid('ulid')->nullable();
            $table->string('title');
        });

        $app['db']->connection()->getSchemaBuilder()->create('comments', function (Blueprint $table) {
            $table->increments('id');
            $table->ulid('custom_ulid')->nullable();
            $table->ulid('other_custom_ulid')->nullable();
            $table->string('comment');
        });
    }
}


tests/SupportStrTest.php
<?php

namespace Vanthao03596\LaravelUlid\Tests;

use Carbon\Carbon;
use Illuminate\Support\Str;
use Ulid\Ulid;

class SupportStrTest extends TestCase
{
    public function testUlid()
    {
        $this->assertInstanceOf(Ulid::class, Str::ulid());
        $datetime = Carbon::now()->subHour();
        $this->assertSame($datetime->getTimestamp() * 1000, Str::ulid(false, $datetime)->toTimestamp());
    }
}


tests/GenerateUlidTest.php
<?php

namespace Vanthao03596\LaravelUlid\Tests;

use Illuminate\Database\Eloquent\Model;
use Vanthao03596\LaravelUlid\GeneratesUlid;

class GenerateUlidTest extends TestCase
{
    public function testSetUlidWhenCreateNewModel()
    {
        $post = Post::create(['title' => 'Test post']);

        $this->assertNotNull($post->ulid);
    }

    public function testDoesNotOverridTheUlidIfItAlreadySet()
    {
        $ulid = '01fdw451ch10nge8b0gkjx553q';

        $post = Post::create(['title' => 'Test post', 'ulid' => $ulid]);

        $this->assertSame($ulid, $post->ulid);
    }

    public function testAllowConfigurableUlidColumnName()
    {
        $comment = Comment::create(['comment' => 'Test comment']);

        $this->assertNotNull($comment->custom_ulid);
    }

    public function testMultipleUlidColumnNames()
    {
        $comment = Comment::create(['comment' => 'Test comment']);

        $this->assertNotNull($comment->custom_ulid);
        $this->assertNotNull($comment->other_custom_ulid);
    }
}

class Post extends Model
{
    use GeneratesUlid;
    protected $guarded = [];

    public $timestamps = false;
}

class Comment extends Model
{
    use GeneratesUlid;
    protected $guarded = [];

    public $timestamps = false;

    public function ulidColumns(): array
    {
        return ['custom_ulid', 'other_custom_ulid'];
    }
}


src/GeneratesUlid.php
<?php

namespace Vanthao03596\LaravelUlid;

use Illuminate\Support\Str;
use Ulid\Ulid;

trait GeneratesUlid
{
    public static function bootGeneratesUlid(): void
    {
        static::creating(function ($model) {
            foreach ($model->ulidColumns() as $item) {
                /* @var \Illuminate\Database\Eloquent\Model|static $model */
                $ulid = Str::ulid();

                if (isset($model->attributes[$item]) && ! is_null($model->attributes[$item])) {
                    /* @var \Ulid\Ulid $ulid */
                    $ulid = Ulid::fromString($model->attributes[$item], true);
                }

                $model->{$item} = (string) $ulid;
            }
        });
    }

    /**
     * The name of the column that should be used for the ULID.
     *
     * @return string
     */
    public function ulidColumn(): string
    {
        return 'ulid';
    }

    /**
     * The names of the columns that should be used for the ULID.
     *
     * @return array
     */
    public function ulidColumns(): array
    {
        return [$this->ulidColumn()];
    }
}


src/SkeletonServiceProvider.php
<?php

namespace VendorName\Skeleton;

use Illuminate\Support\ServiceProvider;
use VendorName\Skeleton\Commands\SkeletonCommand;

class SkeletonServiceProvider extends ServiceProvider
{
    public function boot(): void
    {
        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__ . '/../config/skeleton.php' => config_path('skeleton.php'),
            ], 'config');

            $this->publishes([
                __DIR__ . '/../resources/views' => base_path('resources/views/vendor/skeleton'),
            ], 'views');

            $migrationFileName = 'create_skeleton_table.php';
            if (! $this->migrationFileExists($migrationFileName)) {
                $this->publishes([
                    __DIR__ . "/../database/migrations/{$migrationFileName}.stub" => database_path('migrations/' . date('Y_m_d_His', time()) . '_' . $migrationFileName),
                ], 'migrations');
            }

            $this->commands([
                SkeletonCommand::class,
            ]);
        }
    }

    public function register(): void
    {
        $this->mergeConfigFrom(__DIR__ . '/../config/skeleton.php', 'skeleton');
    }

    public static function migrationFileExists(string $migrationFileName): bool
    {
        $len = strlen($migrationFileName);
        foreach (glob(database_path("migrations/*.php")) as $filename) {
            if ((substr($filename, -$len) === $migrationFileName)) {
                return true;
            }
        }

        return false;
    }
}


src/LaravelUlidServiceProvider.php
<?php

namespace Vanthao03596\LaravelUlid;

use Illuminate\Database\Schema\Blueprint;
use Illuminate\Database\Schema\ColumnDefinition;
use Illuminate\Database\Schema\ForeignIdColumnDefinition;
use Illuminate\Database\Schema\Grammars\Grammar;
use Illuminate\Support\ServiceProvider;
use Illuminate\Support\Str;
use RuntimeException;
use Ulid\Ulid;

class LaravelUlidServiceProvider extends ServiceProvider
{
    public function boot(): void
    {
        Grammar::macro('typeUlid', function () {
            switch (class_basename(static::class)) {
                case 'PostgresGrammar':
                case 'MySqlGrammar':
                    return 'char(26)';

                case 'SQLiteGrammar':
                    return 'varchar';

                case 'SqlServerGrammar':
                    return 'nchar(26)';

                default:
                    throw new RuntimeException('This database is not supported.');
            }
        });

        Blueprint::macro('ulid', function ($column): ColumnDefinition {
            /* @var Blueprint $this */
            return $this->addColumn('ulid', $column);
        });

        Blueprint::macro('foreignUlid', function ($column): ColumnDefinition {
            /* @var Blueprint $this */
            return $this->addColumnDefinition(new ForeignIdColumnDefinition($this, [
                'type' => 'ulid',
                'name' => $column,
            ]));
        });

        Blueprint::macro('ulidMorphs', function ($name, $indexName = null) {
            /* @var Blueprint $this */
            $this->string("{$name}_type");

            $this->ulid("{$name}_id");

            $this->index(["{$name}_type", "{$name}_id"], $indexName);
        });

        Blueprint::macro('nullableUlidMorphs', function ($name, $indexName = null) {
            /* @var Blueprint $this */
            $this->string("{$name}_type")->nullable();

            $this->ulid("{$name}_id")->nullable();

            $this->index(["{$name}_type", "{$name}_id"], $indexName);
        });

        Str::macro('ulid', function (bool $lowercase = true, string $dateTime = null) {
            return $dateTime
                ? Ulid::fromTimestamp(strtotime($dateTime) * 1000, $lowercase)
                : Ulid::generate($lowercase);
        });
    }
}
